### PR TITLE
We should call the UnLoad methods when we detect an error during call…

### PR DESCRIPTION
**Pull request by normanmaurer**
…ing OnLoad

Motivation:

We should ensure we call *UnLoad when we detect an error during calling *OnLoad and previous *OnLoad calls were succesfull.

Modifications:

Correctly call *UnLoad when needed.

Result:

More correct code and no leaks when an error happens during loading the native lib.

### CODE REVIEW

**Comment by normanmaurer** [comment]
This came up here https://github.com/netty/netty/pull/8231#discussion_r213451757.

We should do the same in netty as well.

File: openssl-dynamic/src/main/c/jnilib.c [hunk 1]
```diff
@@ -291,27 +291,41 @@ static const jint method_table_size = sizeof(method_table) / sizeof(method_table
 // JNI Method Registration Table End
 
 static jint netty_internal_tcnative_Library_JNI_OnLoad(JNIEnv* env, const char* packagePrefix) {
+    int errorOnLoadCalled = 0;
```
Comment (carl-mastrangelo) [inline openssl-dynamic/src/main/c/jnilib.c]: for coolness points, you can make these labels and goto the correct one:

```
bufferOnLoadCalled: netty_internal_tcnative_Buffer_JNI_OnUnLoad(env);
errorOnLoadCalled: netty_internal_tcnative_Error_JNI_OnUnLoad(env);

```

Comment (ejona86) [reply openssl-dynamic/src/main/c/jnilib.c]: You "fall through" from one label to the next, so it will call multiple unload methods naturally.

Comment (normanmaurer) [reply openssl-dynamic/src/main/c/jnilib.c]: Ah got it... not sure this is clearer tho. I think I would prefer to keep it as it is

**Pull request merged**